jax/_src/third_party/scipy/special.py

```diff
@@ -0,0 +1,299 @@
+from __future__ import annotations
+
+from functools import partial
+
+import jax.numpy as jnp
+from jax import jit
+
+from jax._src import custom_derivatives
+from jax._src.numpy.util import promote_args_inexact
+from jax._src.typing import Array, ArrayLike
+
+
+@jit
+def sincospisquaredhalf(
+  x: Array,
+) -> tuple[Array, Array]:
+  """
+  Accurate evaluation of sin(pi * x**2 / 2) and cos(pi * x**2 / 2).
+
+  As based on the sinpi and cospi functions from SciPy, see:
+  - https://github.com/scipy/scipy/blob/v1.14.0/scipy/special/special/cephes/trig.h
+  """
+  x = jnp.abs(x)
+  # define s = x % 2, y = x - s, then
+  # r = (x * x / 2) % 2
+  #   = [(y + s)*(y + s)/2] % 2
+  #   = [y*y/2 + s*y + s*s/2] % 2
+  #   = [(y*y/2)%2 + (s*y + s*s/2)%2]%2
+  #   = [0 + (s*(y+s/2))%2]%2
+  #   = [s*(x-s/2)]%2
+  s = jnp.fmod(x, 2.0)
+  r = jnp.fmod(s * (x - s / 2), 2.0)
+
+  sinpi = jnp.where(
+    r < 0.5,
+    jnp.sin(jnp.pi * r),
+    jnp.where(
+      r > 1.5,
+      jnp.sin(jnp.pi * (r - 2.0)),
+      -jnp.sin(jnp.pi * (r - 1.0)),
+    ),
+  )
+  cospi = jnp.where(
+    r == 0.5,
+    0.0,
+    jnp.where(r < 1.0, -jnp.sin(jnp.pi * (r - 0.5)), jnp.sin(jnp.pi * (r - 1.5))),
+  )
+
+  return sinpi, cospi
+
+
+@custom_derivatives.custom_jvp
+def fresnel(x: ArrayLike) -> tuple[Array, Array]:
+  r"""The Fresnel integrals
+
+  JAX implementation of :obj:`scipy.special.fresnel`.
+
+  The Fresnel integrals are defined as
+    .. math::
+       S(x) &= \int_0^x \sin(\pi t^2 /2) dt \\
+       C(x) &= \int_0^x \cos(\pi t^2 /2) dt.
+
+  Args:
+    x: arraylike, real-valued.
+
+  Returns:
+    Arrays containing the values of the Fresnel integrals.
+
+  Notes:
+     The JAX version only supports real-valued inputs, and
+     is based on the SciPy C++ implementation, see
+     `here
+     <https://github.com/scipy/scipy/blob/v1.14.0/scipy/special/special/cephes/fresnl.h>`_.
+     For ``float32`` dtypes, the implementation is directly based
+     on the Cephes implementation ``fresnlf``.
+
+     As for the original Cephes implementation, the accuracy
+     is only guaranteed in the domain [-10, 10]. Outside of
+     that domain, one could observe divergence between the
+     theoretical derivatives and the custom JVP implementation,
+     especially for large input values.
+  """
+  if jnp.iscomplexobj(x):
+    raise NotImplementedError(
+        'Support for complex-valued inputs is not implemented yet.')
+
+  xxa, = promote_args_inexact("fresnel", x)
+
+  # This part is mostly a direct translation of SciPy's C++ code.
+
+  single_precision = (xxa.dtype == jnp.float32)
```

Comment:
We need to decide what to do with float16 and bfloat16 inputs. As currently written, the implementation will effectively promote them to whatever the widest available float type is (float64 if enable_x64 is True, float32 if not). Generally we would want the output dtype to match the input dtype

Reply:
Take a look at how `bessel_i0e_impl` handles various input dtypes – I think we should use the same pattern here.

Reply:
> We need to decide what to do with float16 and bfloat16 inputs. As currently written, the implementation will as a side-effect promote them to whatever the widest available float type is (float64 if enable_x64 is True, float32 if not). Generally we would want the output dtype to match the input dtype

I totally forgot they existed, good point. Well, upcasting is the easy solution (as what they do with `bessel_i0e_impl`). But, before that, I think it would be important to check if using, e.g., the `float32` coefficients for `float16` results in accurate results, or not.

Reply:
After some tests, it appears that `float16` overwflows for any value in the "other" range, which means the values are NaNs for x > 3 (roughly). For `bfloat16`, series expansions start to diverge for x > 8~9, which is better but still not very good.

Based on the Bessel example, I decided to manually upcast the arrays elements to `float32`, and downcast them if needed.